plugins/docker/stats_handling.go

```diff
@@ -0,0 +1,361 @@
+	go m.handleDockerEvents(stopChan)
+	m.ir.LogMessage("Listen ")
+
+	source := new(Source)
+	events := make(chan *StatsAttachEvent)
+	m.addStatsListener(events)
+	defer m.removeStatsListener(events)
+	//Add listeners for all currently running containers
+	for {
+		select {
+		case event := <-events:
+			m.ir.LogMessage("Event.ID : " + event.ID + " , event.Type : "+ event.Type + " , event.Name : "+  event.Name)
+
+			if event.Type == "attach" && (source.AllContainers() ||
+				(source.ID != "" && strings.HasPrefix(event.ID, source.ID)) ||
+				(source.Name != "" && event.Name == source.Name) ||
+				(source.Filter != "" && strings.Contains(event.Name, source.Filter))) {
+
+				pump := m.Get(event.ID)
+				pump.AddStatsListener(statsstream)
+				defer func() {
+					if pump != nil {
+						pump.RemoveStatsListener(statsstream)
+					}
+				}()
+			} else if source.ID != "" && event.Type == "detach" &&
+				strings.HasPrefix(event.ID, source.ID) {
+				return
+			}
+		case <-closer:
+			return
+		}
+	}
+}
+
+// Attach to all running containers
+func (m *StatsManager) attachAll() error {
+	containers, err := m.client.ListContainers(docker.ListContainersOptions{})
+	if err != nil {
+		m.ir.LogError(err)
+		return err
+	}
+
+	for _, listing := range containers {
+		err := withRetries(func() error {
+			return m.statsAttach(listing.ID[:12], m.client)
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+
+	m.ir.LogMessage("Stats Attached to all containers")
+
+	return nil
+}
+
+// Attach to the output of a single running container
+func (m *StatsManager) statsAttach(id string, client DockerClient) error {
+	m.ir.LogMessage(fmt.Sprintf("Attaching container: %s", id))
+
+	done := make(chan bool)
+	statsrd := make(chan *docker.Stats)
+	failure := make(chan error)
+
+	// Spin up one of these for each container we're watching
+	go func() {
+		// This will block until the container exits
+		err := client.Stats(docker.StatsOptions{
+			ID:    	id,
+			Stats:  statsrd,
+			Stream: true,
+			Done:   done,
+		})
+
+		// Once it has exited, close our pipes
+		close(statsrd)
+		close(done)
+		if err != nil {
+			close(done)
+			failure <- err
+		}
+	}()
+
+	container, err := m.client.InspectContainer(id)
+	if err != nil {
+		m.errors <- err
+	}
+	name := container.Name[1:]
+	fields, err := extractFields(id, m.client, m.fieldsFromLabels, m.fieldsFromEnv, m.nameFromEnv)
+	if m.nameFromEnv != "" {
+		if alt_name, ok := fields[m.nameFromEnv]; ok && alt_name != "" {
+			name = alt_name
+		}
+	}
+	fields["ContainerID"] = id
+	fields["ContainerName"] = name
+	var resultStats []*docker.Stats
+	for {
+		stats, ok := <-statsrd
+		if !ok {
+			break
+		}
+		resultStats = append(resultStats, stats)
+		m.Lock()
+		m.attached[id] = NewStatsPump(statsrd, name, fields)
+		m.Unlock()
+		m.send(&StatsAttachEvent{ID: id, Name: name, Type: "attach"})
+		return nil
+	}
+	return nil
+}
+
+func (m *StatsManager) send(event *StatsAttachEvent) {
+	m.RLock()
+	defer m.RUnlock()
```

Comment:
This might be another small nit, but the `defer`s in the next few methods feel like overkill. In many cases the defer is only one line away from where a non-deferred call would be placed.